scalar: let more functional tests pass

dscho opened the pull request:
A couple assorted fixes, most notably disabling the untracked cache on Windows to work around test failures.

dscho committed:
fixup! scalar unregister: handle deleted enlistment directory gracefully

We will introduce this test much earlier than even the `clone` test.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: implement the `clone` subcommand

Let's rename the variable to reflect what it is about: the path of the
enlistment's top-level directory.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! fixup! scalar: do initialize `gvfs.sharedCache`

The `root` variable was just renamed.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: implement the `clone` subcommand

Let's disallow clonining into existing directories. It simplifies the
code, and the upcoming fix where we need to roll back the `git init` if
we're detecting that the local cache root would be inside the `src`
directory.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar clone: fix `dir_inside_of()` on case-insensitive filesystems

The functional tests actually expect the enlistment _not_ to exist if
the "is cache root inside of src/" check prevents `scalar clone` from
completing.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar register: set recommended config settings

Let's drop the `core.filemode` setting: `git init` already initializes
this correctly.

And then, disable the untracked cache on Windows. This hurts a bit
because the performance improvement is notable, but we have to turn it
off because it is apparently unreliable: the functional tests
consistently fail with `core.untrackedCache = true`.

Signed-off-by: Johannes Schindelin <johasc@microsoft.com>

dscho committed:
fixup! scalar unregister: handle deleted enlistment directory gracefully

The name `vanish` is a much better name for this directory.

Signed-off-by: Johannes Schindelin <johasc@microsoft.com>

mjcheetham commented on contrib/scalar/scalar.c:
```diff
@@ -158,12 +133,10 @@ static int set_recommended_config(void)
-		/*
-		 * Windows-specific settings.
-		 */
+#ifndef WIN32
 		{ "core.untrackedCache", "true" },
-		{ "core.filemode", "true" },
+#else
+		{ "core.untrackedCache", "false" },
 #endif
```
nit: would it not be more "natural" to invert this `#ifndef` to an `#ifdef`?

"if windows then X else Y"

dscho replied:
In this instance, I thought that we'd use the desirable arm first: "_unless_ this is Windows, we use the untracked cache".

I should probably add a comment to clarify that this is unfortunate and that we'll want to spend some quality time trying to make untracked cache work on Windows and get rid of the `#else`.

dscho replied:
Did that: https://github.com/microsoft/git/pull/350/commits/94d7b05b4c8a187f17358d3d076a49846d61cbcd#diff-f9053a8d79e6d409ddf144989d8202388ac5edadbdc3b6140a6b94c894110950R140-R149

dscho committed:
git_config_set_multivar_in_file_gently(): add a lock timeout

In particular when multiple processes want to write to the config
simultaneously, it would come in handy to not fail immediately when
another process locked the config, but to gently try again.

This will help with Scalar's functional test suite which wants to
register multiple repositories for maintenance semi-simultaneously.

Signed-off-by: Johannes Schindelin <johasc@microsoft.com>

dscho committed:
scalar: set the config write-lock timeout to 150ms

By default, Git fails immediately when locking a config file for writing
fails due to an existing lock. With this change, Scalar-registered
repositories will fall back to trying a couple times within a 150ms
timeout.

Signed-off-by: Johannes Schindelin <johasc@microsoft.com>
